### PR TITLE
[precompile] Neutralize ambient autocast in served artifacts

### PR DESCRIPTION
**Pull request by bobrenjc93**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.17.0) (oldest at bottom):
* #197353
* #197352
* #197347
* #197345
* #197344
* #197456
* #197763
* #197767
* #197766
* #197343
* #197351
* #197839
* #197998
* #197849
* #197844
* #197689
* #197720
* #197600
* #197388
* #197387
* __->__ #197455
* #197341
* #197342
* #197348
* #197349

**Summary.** A precompile artifact already has the capture's casts baked in -- ATen casts for the make_fx driver, generated kernels for inductor -- but serving still re-dispatches, so a process that has a `torch.autocast` region on casts a second time and the served call comes back in a different dtype than the capture produced. This commit closes that inside the emitted drivers: both wrap their `call(...)` in `torch._C._DisableAutocast`, the context manager bound in `torch/csrc/autograd/init.cpp` that holds a single `c10::impl::ExcludeDispatchKeyGuard` over `c10::autocast_dispatch_keyset`, so every autocast dispatch key the serving build has is excluded for the duration of a served call. One dispatch-key guard rather than a Python walk over `torch._C._autocast_supported_devices()` clearing per-device TLS bits, and it is strictly wider rather than merely shorter: `at::autocast::is_autocast_enabled(device)` IS `!tls_is_dispatch_key_excluded(key)`, so excluding the whole keyset (all ten keys, `AutocastMPS` and `AutocastMAIA` included) covers the supported devices this build ships no `torch.<device>` module for -- which a `hasattr(torch, device)`-guarded loop drops silently -- with no probe, no constructor that can raise, and no per-device skip to report. It is also the mechanism already in the tree for exactly this job: AOTAutograd, the subsystem precompile lowers through, emits this same guard into its own generated runtime source (`torch/_functorch/_aot_autograd/runtime_wrappers.py` binds `_DisableAutocast_` and emits `with _DisableAutocast_():` around generated calls) under the same "the casts are baked in, do not re-dispatch under autocast" reasoning, and invariant 7's self-containment still holds because `torch._C._DisableAutocast` is an attribute path the artifact carries as TEXT and resolves off its own `import torch as _torch` -- like every other `_torch.` name the driver already uses -- not an import of a private Python helper whose body could skew with the torch version. Both entry points emit the same driver, so this is a caller-visible behavior change for every artifact written from here on -- including by the already-public `torch.compiler.precompile` -- and the contract is therefore stated in `_PrecompileApi.__call__`'s and `_PrecompileApi.load`'s docstrings, which is what `help()` renders, and in `docs/source/torch.compiler_api.md`, which is what Sphinx renders for `torch.compiler.precompile`, as well as in `capture()`'s docstring, in the emitted artifact's own header and in Note [precompile programming model].

## The guard

```python
with _torch._C._DisableAutocast(), _torch.no_grad():   # _eager_forward
    out = list(call([*pb, *user_flat]))
```

```python
        with _torch._C._DisableAutocast():             # _inductor_forward
            out = list(call([*pb, *user_flat]))
```

That is the whole mechanism in both drivers. Autocast state is per device type at the Python level -- `torch.amp.autocast(device, enabled=False)` covers one device -- which invites a loop over `torch._C._autocast_supported_devices()`, and `torch._functorch._aot_autograd.graph_capture_wrappers.disable_autocast` is such a loop. But the per-device Python bit and the dispatch key are the same state seen from two ends (`is_autocast_enabled` reads the exclusion set), so excluding the keyset once subsumes the loop: it needs no `hasattr(torch, device)` guard and therefore drops none of the supported devices this build has no module for, it cannot raise out of an `autocast` constructor and so has no skip path to report or document, and there is no gate to reason about because excluding an already-excluded key costs nothing. The inductor driver needs the guard even though inductor compiled the graph, because the generated code calls `extern_kernels` for everything it did not fuse and those re-dispatch through the autocast key; in the eager driver the guard composes with the existing `no_grad` on one `with`.

The narrower alternative -- record the captured graph's device types in the artifact and neutralize only those -- is rejected: an op whose schema and fake impl are cpu->cpu but whose body moves work to another device is not in such a tag, and the tag would add a metadata field plus a compatibility branch that the keyset guard has nothing to version.

## Where the contract ships

`capture()` is not exported yet, and Note [precompile programming model] is a `#` comment in a private module, so neither is reachable from a user of the shipped API. Two places are, and they are not the same place. `precompile.__doc__ = _PrecompileApi.__call__.__doc__` makes `__call__`'s docstring what `help(torch.compiler.precompile)` prints, so the paragraph goes there and a short version goes on `_PrecompileApi.load`. Sphinx reads neither: `precompile` is a callable instance, so it is deliberately omitted from `docs/source/torch.compiler_api.md`'s `autosummary` block and documented by hand there in `.. py:function:: precompile` / `.. py:method:: precompile.load` directives, which render their own body and never consult `__doc__`. That page is the rendered public reference, so the contract ships there too -- one sentence in the same `.. note::` that already carries the sibling specialization contract (each input's dtype and device are specialized, and the inductor backend additionally specializes on input memory format), which until now told a reader only that INPUT dtypes are pinned and nothing about the output side moving under an ambient region, plus one sentence in the `precompile.load` block. The generated artifact's own headers (`_GENERATED_HEADER`, `_EAGER_GENERATED_HEADER`) carry it as well, since a reader who opens the emitted file meets that text and no docstring. All of them say the same two things, and only those: a served call returns the capture's dtypes, NOT the dtypes the same eager call returns inside an ambient autocast region; and the remedy is to capture under the autocast you want baked in. Nothing about the mechanism is stated in the rendered reference, since the guard's coverage is total and there is no residual case for a caller to work around.

Test Plan:

The served output under `torch.autocast("cpu", dtype=torch.bfloat16)` equals the float32 capture-time result: on eager for `relu(Linear(x))`, whose inductor lowering calls `extern_kernels.addmm(..., out=)`, an overload with no CPU autocast registration and so one that cannot exercise the neutralization; and on inductor for an `nn.LSTM`, whose lowering keeps `mkldnn_rnn_layer.default` as an extern kernel that ambient autocast does reach -- that test asserts its own premise (`mkldnn_rnn_layer` really is in the artifact) and is skipped unless mkldnn is both built AND enabled, since the decomposition that keeps the kernel gates on the runtime `torch._C._get_mkldnn_enabled()` flag rather than on `_has_mkldnn`, and with mkldnn built but disabled the LSTM lowers to `one_layer_lstm` and the premise assertion would fail rather than skip. Because that arm is the only one that can OBSERVE the inductor driver's guard, the CPU test additionally recaptures with `backend="inductor"` and asserts `_DisableAutocast` in the artifact text, which pins the inductor half on every build. On CUDA, `relu(Linear(x))` served inside `torch.autocast("cuda", dtype=torch.float16)` comes back float32 where the eager `model(x)` comes back float16; eager only, because the CUDA arm has the same `extern_kernels.addmm(..., out=)` premise problem as the CPU one and an inductor arm there would pass whether or not the guard is present. The documented remedy has its own round trip: a capture made INSIDE `torch.autocast("cpu", dtype=torch.bfloat16)` serves the capture-time bfloat16 result (value and dtype, not dtype alone) outside every region, AND inside an ambient `float16` region -- the only coverage of a bf16-baked artifact served under a different dtype, which is the direction where a second cast silently downgrades precision instead of upgrading it, and the arm that makes that test discriminate at all (without it the served call sits outside every region and the guard has nothing to do). Nothing tests the guard's internals, because it has none: there is no device list to mock, no constructor to make refuse, and no unwind to arrange, so the behavior tests above are the whole surface.

```
python test/test_precompile.py
python test/functorch/test_compile_to_python.py
lintrunner -a torch/_precompile.py torch/_precompile_driver.py test/test_precompile.py docs/source/torch.compiler_api.md
```

The mutation that matters is removing the guard from a driver. Removing the eager one fails the dtype assertion in the CPU, CUDA and bake-in tests; removing the inductor one fails the artifact-text assertion on any build and, where mkldnn is enabled, also errors inside oneDNN on the mixed-dtype LSTM primitive. A rename of the guard is caught the same way, and additionally by construction on any path that runs, since the artifact is `exec`'d from its own text on every `load`. The full suite otherwise matches the pre-existing baseline in this environment, which has a CUDA device but no triton: the remaining errors are the CUDA numerics tests and the two autotune-cache tests.

Authored with an AI assistant.

### CODE REVIEW

**Comment by bobrenjc93** [comment]
@claude

**Comment by bobrenjc93** [comment]
Thanks -- every item below is dispositioned against head c4ca7852adf.

> `torch._C._DisableAutocast` already exists and is what AOTAutograd emits for this exact case.

**Fixed** in c4ca7852adf: both drivers now enter `torch._C._DisableAutocast()` (the eager one together with `no_grad`), and `_autocast_off`, the skip set, the emit-side literal and the metadata-comment churn are gone. The message now argues the mechanism the way the review does: one dispatch-key guard over all ten autocast keys, the same guard AOTAutograd bakes into its generated runtime source, an attribute path the artifact text already carries.

> The `hasattr(torch, dev)` guard silently drops four of the ten supported devices, contradicting the documented contract.

**Fixed** in c4ca7852adf: no per-device loop remains; the docstrings and the docs page state only the contract (a served call ignores the serving process's ambient autocast because the casts are baked in).

> the `is_autocast_enabled` gate is not the no-op the docstring claims.
> the docstring overstates why the `AssertionError` catch is needed.
> The comment volume is disproportionate.

**Fixed** in c4ca7852adf: all three describe code that no longer exists; the prose went with it.

> `test_a_device_the_captured_graph_never_names_is_neutralized_too` asserts the opposite of its own premise.
> `test_autocast_off_unwinds_a_failure_partway_through` silently depends on device ordering
> `test_a_supported_device_this_build_has_no_module_for_is_ignored` does not test what its name says.
> No test that the emitted `_AUTOCAST_SKIPS_REPORTED` name matches the declared one.
> Global-state cleanup restores a hardcoded value rather than the prior one.

**Fixed** in c4ca7852adf: those tests described `_autocast_off` internals and are deleted; the kept tests pin the behaviour (served output ignores an ambient CPU autocast region; the CUDA test where triton is available) and no kept test mutates global autocast state. Removing the guard from both drivers fails every kept test.

> `test_served_output_ignores_ambient_cuda_autocast`'s `backend="inductor"` arm is likely vacuous

**Fixed** in c4ca7852adf: the inductor arm is dropped, for the reason the review gives (the lowering's `out=` addmm has no autocast registration).

> No coverage of the train-step/grad path under ambient autocast

**Declined**: the driver accumulates grads with `p.grad.add_(g)`, so a bf16 gradient added into an fp32 `.grad` stays fp32 and the assertion cannot discriminate; a discriminating test needs a fresh fixture, and the forward tests already pin the guard both drivers share.

> puts an implementation caveat in the rendered public reference.

**Fixed** in c4ca7852adf: the caveat is gone with the mechanism.

**Comment by bobrenjc93** [comment]
@claude please re-review at head c4ca7852adf: the per-device loop is replaced by torch._C._DisableAutocast as you proposed (dispositions in the comment above), and the stack was rebased onto current main.

**Comment by bobrenjc93** [comment]
Thanks -- every item below is dispositioned against head fb92035ee56.

> The `_inductor_forward` guard (`torch/_precompile_driver.py:320`) has zero coverage on a build without mkldnn.

**Fixed** in fb92035ee56: the ungated CPU autocast test recaptures with `backend="inductor"` and asserts `_DisableAutocast` in the emitted artifact text; an artifact captured outside autocast carries exactly one such guard (AOTAutograd's is gated on compile-time autocast), and removing the driver's guard fails the assertion.

> `test/test_precompile.py:3147-3148` gates on the wrong mkldnn predicate

**Fixed** in fb92035ee56: the skip gate requires both `torch.backends.mkldnn.is_available()` and `torch._C._get_mkldnn_enabled()`.

> The train-step declination is contradicted by the driver code.

**Declined** on cost, and you are right about the reasoning: on a fresh model the first served call assigns `p.grad`, whose setter checks the grad dtype against the parameter's, and the train-step fixture returns `out`, so a dtype assertion would discriminate. The forward tests pin the guard both drivers share; the reply above was wrong about why.

> Two minor test points.

**Fixed** in part in fb92035ee56: the extern-kernel comment now says what the test asserts and that the second cast fails inside oneDNN on this primitive. The outside-region `served(...)` line stays: it is the only assertion of the "serves the captured dtypes outside every region" half the test's comment claims.

> `torch/_precompile.py:272-276` argues against an alternative that was never in the tree.

**Fixed** in fb92035ee56: the Note states the mechanism only; the description keeps one sentence of justification.

**Comment by bobrenjc93** [comment]
@claude please re-review at head fb92035ee56 (round-3 dispositions above: build-independent assertion of the inductor guard, the built-and-enabled mkldnn gate, Note trimmed to the mechanism).

**Comment by bobrenjc93** [comment]
Thanks for the approval -- the nits are dispositioned against head fb92035ee56.

> The new artifact-text assertion is satisfiable by AOTAutograd's own emission, so it is not the rename/deletion backstop it reads as.

**Accepted**, folded for the next push that touches this PR: the assertion matches the driver's emitted line, `with _torch._C._DisableAutocast():`, which only `inspect.getsource` of the driver can produce, with a comment saying why AOTAutograd's own guard is absent here.

> `test_served_output_ignores_ambient_autocast` now does two unrelated things under one name.
> `test_capturing_under_autocast_bakes_the_casts_in` only covers `backend="eager"`.

**Declined**: test organisation and an extra inductor arm add lines for no new behaviour; the emission check sits with the test that already captures on both backends.

> `docs/source/torch.compiler_api.md:84-88` and `_precompile.py:3132-3141` promise behavior an already-written artifact does not have.

**Declined**: the API is a prototype with no compatibility promise for artifacts written by earlier builds, as the page-level warning states; version-scoping each sentence would restate that warning per feature.

**Comment by bobrenjc93** [comment]
Closing: the author is taking a different direction with this work. Nothing here landed; the branch is left in place.
